### ISSUE TITLE
API endpoint for obtaining subsumed annotations

### ISSUE THREAD

**Issue by hlapp** [phenoscape/phenoscape-kb-services#146]
It seems there is currently no API endpoint for obtaining the subsumed annotations for a given subsumer, or for a list of subsumers.

(There is `/similarity/subsumed_annotations`, which, despite the name, does not seem to provide this function; and there is `/phenotype/query`, which based on signature doesn't seem appropriate, and also queries _phenotypes_ rather than phenotype _annotations_.)

The number of subsumed annotations is the basis for calculating the weights vector for the subsumers in the subsumer matrix, which in turn is the basis for weight-relying similarity metrics such as any IC-based metric. So having access to that through the API would be highly desirable.

**Comment by hlapp** [comment]
Presumably, there would need to be corpus parameter as well, so one can choose the corpus in which to count subsumed annotations.

I understand that currently the database calculates IC scores not based on subsumed annotations, but subsumed phenotypes, or even number of taxa or genes with subsumed annotations. However, none of these numbers seem to have an endpoint that would make them accessible through the API.

**Comment by balhoff** [comment]
Am I right that you would also want the corpus size for any corpus exposed in this service?

**Comment by hlapp** [comment]
Yes, but doesn't the "/similarity/corpus_size" endpoint provide this already? But arguably the identifier of the corpus should probably be a parameter.

**Comment by balhoff** [comment]
> but doesn't the "/similarity/corpus_size" endpoint provide this already?

@hlapp yes that's correct.

**Comment by balhoff** [comment]
Should this be a JSON result format, or do you prefer to have this again as CSV?

```csv
http://example.org/term1,42
http://example.org/term2,7
```

**Comment by hlapp** [comment]
Either way works fine. JSON that is not a nested structure and repeats the same elements is very straightforward to use. CSV on the other hand is presumably is less verbose, i.e., at least faster to transmit?

**Comment by balhoff** [comment]
This is implemented. The only supported "corpus" at the moment is `http://kb.phenoscape.org/sim/taxa` (taxon variation profiles). I think I'll close this issue, but we can open additional issues to handle other kinds of frequency counts or return formats.

**Comment by balhoff** [comment]
The `http://kb.phenoscape.org/sim/taxa` corpus expects identifiers for *phenotypes*, btw.

**Comment by hlapp** [comment]
> The `http://kb.phenoscape.org/sim/taxa` corpus expects identifiers for _phenotypes_, btw.

Not sure what this means or why. As stated in the beginning, this is needed for the subsumer terms, i.e., those returned by the subsumer matrix endpoint. I don't know whether that endpoint only returns phenotypes as subsumers if only phenotypes are in the input list (I doubt it). However, given that the subsumer matrix accepts any (anatomy) terms, and assuming that phenotype terms aren't (or at least aren't the only) subsumers of anatomy entity terms, this would mean that the subsumer matrix will have rows that aren't phenotypes.

I'm going to open this issue again, because until the endpoint returns a term frequency at least for any anatomy term potentially returned by the subsumer matrix endpoint, it is not useful for the original purpose.

**Comment by balhoff** [comment]
> I don't know whether that endpoint only returns phenotypes as subsumers if only phenotypes are in the input list (I doubt it).

Yes this is true.

We have to define how the terms being submitted relate to some kind of instance being counted (e.g. taxon variation profile). So a corpus identifier is a specification of both what things to count, and how the submitted terms relate to those things. So the KB knows that taxon variation profiles are related to a set of phenotypes with a particular relation, and there are subsumer classes in the KB with the same logical shape as those phenotypes.

The subsumer matrix accepts any terms at all, so it all depends on what you start with. If you have some phenotype classes that you got from a query, you can put them in the subsumer matrix service, and then pass the results to this service using the `http://kb.phenoscape.org/sim/taxa` corpus. But there can't be a version of this that automatically handles any kind of input term; they are in various different hierarchies.

We could define another corpus IRI that returns taxon variation profile counts for entity inputs, if that's what you want. If this doesn't make sense, maybe we can have a phone call. There are a few dimensions to this that I'm hoping we can come up with an ergonomic interface for.

**Comment by hlapp** [comment]
So I think what I'm interested in here in the most basic form are term frequencies as a proxy for probability of encountering a term (which in turn is the basis for IC scores). I.e., given a term, how often does it occur in a given corpus.

Indeed there are different ways to count that frequency. The most basic corpora and counts for IC score calculation for a set of terms would, I think, be the following:
- The number of taxon annotations subsumed by a given term. _(This will inherently be biased for taxa that are more frequently included in studies, because those studies will be likely to include similar or even the same characters, inflating the number of annotations for those characters, and conversely lowering IC scores for the entities and phenotypes used in those annotations. This need not be a bad thing; some taxa, and thus their most distinctive characters, being studies much more frequently than others is a reality, and one might thus want this to be reflected in the IC scores.)_
- The number of taxa with annotations subsumed by a given term. _(Unlike the above, this would remove the effect of some taxa, and thus their most distinctive characters, being much more frequently included in studies than others.)_
- The same for genes and gene annotations.

How's that sound?


For calculating IC score for terms that aren't taxon profiles, the taxon variation profiles don't matter for determining term frequency.

**Comment by balhoff** [comment]
Discussion on 2019-11-12:

Precompute 4 counts for each term
- taxon annotations
- gene annotations
- taxa
- genes

How to count will depend on the type of term (how that hierarchy relates to the phenotype hierarchy).
